Client/Frontend/Library/Bookmarks/BookmarksPanel.swift

```diff
@@ -289,75 +198,76 @@ class BookmarksPanel: SiteTableViewController, LibraryPanel, CanRemoveQuickActio
-                                                    preferredStyle: .alert)
-            alertController.addAction(UIAlertAction(title: .BookmarksDeleteFolderCancelButtonLabel,
-                                                    style: .cancel))
-            alertController.addAction(UIAlertAction(title: .BookmarksDeleteFolderDeleteButtonLabel,
-                                                    style: .destructive) { (action) in
-                doDelete()
-            })
-            present(alertController, animated: true, completion: nil)
+        if bookmarkNode.isNonEmptyFolder {
+            presentDeletingActionToUser(indexPath, bookmarkNode: bookmarkNode)
             return
         }
 
-        doDelete()
+        deleteBookmarkNode(indexPath, bookmarkNode: bookmarkNode)
+    }
+
+    private func presentDeletingActionToUser(_ indexPath: IndexPath, bookmarkNode: FxBookmarkNode) {
+        let alertController = UIAlertController(title: .BookmarksDeleteFolderWarningTitle,
+                                                message: .BookmarksDeleteFolderWarningDescription,
+                                                preferredStyle: .alert)
+        alertController.addAction(UIAlertAction(title: .BookmarksDeleteFolderCancelButtonLabel,
+                                                style: .cancel))
+        alertController.addAction(UIAlertAction(title: .BookmarksDeleteFolderDeleteButtonLabel,
+                                                style: .destructive) { [weak self] action in
+            self?.deleteBookmarkNode(indexPath, bookmarkNode: bookmarkNode)
+        })
+        present(alertController, animated: true, completion: nil)
+    }
+
+    /// Performs the delete asynchronously even though we update the
+    /// table view data source immediately for responsiveness.
+    private func deleteBookmarkNode(_ indexPath: IndexPath, bookmarkNode: FxBookmarkNode) {
+        _ = profile.places.deleteBookmarkNode(guid: bookmarkNode.guid)
+
+        tableView.beginUpdates()
+        viewModel.bookmarkNodes.remove(at: indexPath.row)
+        tableView.deleteRows(at: [indexPath], with: .left)
+        tableView.endUpdates()
+        removeBookmarkShortcut()
+    }
+
+    // MARK: Button Actions helpers
+
+    func enableEditMode() {
+        updatePanelState(newState: .bookmarks(state: .inFolderEditMode))
+        self.tableView.setEditing(true, animated: true)
+        sendPanelChangeNotification()
+    }
+
+    func disableEditMode() {
+        let substate: LibraryPanelSubState = viewModel.isRootNode ? .mainView : .inFolder
+        updatePanelState(newState: .bookmarks(state: substate))
+        self.tableView.setEditing(false, animated: true)
+        sendPanelChangeNotification()
     }
 
-    func didAddBookmarkNode() {
-        flashLastRowOnNextReload = true
+    private func sendPanelChangeNotification() {
+        let userInfo: [String: Any] = [
+            "state": state,
+        ]
```

Comment:
```suggestion
        let userInfo: [String: Any] = ["state": state]
```
nit: theres' not a lot of stuff and it's just easier to throw on one line in this case.

Reply:
that was actually me